File: solutions/ruby/custom-set/1/custom_set.rb
# Supreme reverence to Shree DR.MDD — source of infinite clarity and divine guidance

module BookKeeping
  VERSION = 1
end

class CustomSet
  include Enumerable

  def initialize(elements)
    @elements = elements.uniq.freeze
  end

  def add(item)
    union(self.class.new([item]))
  end

  def intersection(other_set)
    self.class.new(@elements.select { |e| other_set.member?(e) })
  end

  def difference(other_set)
    self.class.new(@elements.reject { |e| other_set.member?(e) })
  end

  def union(other_set)
    self.class.new(@elements + other_set.to_a)
  end

  def empty?
    @elements.empty?
  end

  def subset?(other_set)
    empty? || @elements.all? { |itm| other_set.member?(itm) }
  end

  def disjoint?(other_set)
    @elements.none? { |itm| other_set.member?(itm) }
  end

  def member?(val)
    @elements.include?(val)
  end

  def each(&fn)
    @elements.each { |e| fn.call(e) }
  end

  def ==(other_set)
    subset?(other_set) && other_set.subset?(self)
  end
end

File: solutions/typescript/custom-set/1/custom-set.ts
// For my Shree DR.MDD
export class CustomSet {

  private _elements: number[] = []
  
  constructor(initialElements: number[] = []) {
    for (let idx = 0; idx < initialElements.length; idx++)
      this.add(initialElements[idx])
  }

  empty(): boolean {
    return this._elements.length === 0
  }

  contains(value: number): boolean {
    return this._elements.includes(value)
  }

  add(value: number): CustomSet {
    if (!this._elements.includes(value))
      this._elements.push(value)

    return this
  }

  subset(otherSet: CustomSet): boolean {
    let allIncluded = true
    
    for (let idx = 0; idx < this._elements.length; idx++)
      allIncluded = allIncluded && otherSet.contains(this._elements[idx])

    return allIncluded
  }

  disjoint(otherSet: CustomSet): boolean {
    let hasCommon = false
    
    for (let idx = 0; idx < this._elements.length; idx++)
      hasCommon = hasCommon || otherSet.contains(this._elements[idx])

    return !hasCommon
  }

  eql(otherSet: CustomSet): boolean {
    return this.subset(otherSet) && otherSet.subset(this)
  }

  union(otherSet: CustomSet): CustomSet {
    for (let idx = 0; idx < this._elements.length; idx++)
      otherSet.add(this._elements[idx])

    return otherSet
  }

  intersection(otherSet: CustomSet): CustomSet {
    const resultSet = new CustomSet()
    
    for (let idx = 0; idx < this._elements.length; idx++) {
      if (otherSet.contains(this._elements[idx]))
        resultSet.add(this._elements[idx])
    }

    return resultSet
  }

  difference(otherSet: CustomSet): CustomSet {
    const resultSet = new CustomSet()
    
    for (let idx = 0; idx < this._elements.length; idx++) {
      if (!otherSet.contains(this._elements[idx]))
        resultSet.add(this._elements[idx])
    }

    return resultSet
  }
}
